compiler/hash-ast/src/visualise.rs

```diff
@@ -131,133 +223,676 @@ impl NodeDisplay for Literal {
+    }
+}
+
+impl NodeDisplay for FunctionDefArg {
+    fn node_display(&self) -> Vec<String> {
+        // deal with the name and type as seperate branches
+        let mut arg_components = vec![vec![format!("name {}", self.name.node_display().join(""))]];
+
+        if let Some(ty) = &self.ty {
+            arg_components.push(ty.node_display())
+        }
+
+        let component_lines = draw_branches_for_children(&arg_components);
+        iter::once("arg".to_string())
             .into_iter()
-            .map(|line| pad_str(line.as_str(), ' ', indent, Alignment::Left))
+            .chain(component_lines)
+            .collect()
+    }
+}
+
+impl NodeDisplay for StructLiteralEntry {
+    fn node_display(&self) -> Vec<String> {
+        let name = self.name.node_display();
+
+        let value = iter::once("value".to_string())
+            .chain(child_branch(&self.value.node_display()))
             .collect();
 
-        lines.extend(next_lines);
-        lines
+        iter::once("entry".to_string())
+            .chain(draw_branches_for_children(&[name, value]))
+            .collect()
+    }
+}
+
+impl NodeDisplay for AccessName {
+    fn node_display(&self) -> Vec<String> {
+        let names: Vec<&str> = self.names.iter().map(|n| n.body.string.as_ref()).collect();
+        vec![format!("\"{}\"", names.join("::"))]
+    }
+}
+
+impl NodeDisplay for Name {
+    fn node_display(&self) -> Vec<String> {
+        vec![format!("\"{}\"", self.string)]
     }
 }
 
 impl NodeDisplay for Statement {
-    fn node_display(&self, indent: usize) -> Vec<String> {
-        let mut lines = vec![];
-        let mut next_lines = vec![];
-        let next_indent = indent + 2;
+    fn node_display(&self) -> Vec<String> {
+        let node_name = match &self {
+            Statement::Expr(_) => None,
+            Statement::Return(_) => Some("ret".to_string()),
+            Statement::Block(_) => Some("block".to_string()),
+            Statement::Break => Some("break".to_string()),
+            Statement::Continue => Some("continue".to_string()),
+            Statement::Let(_) => Some("let".to_string()),
+            Statement::Assign(_) => Some("assign".to_string()),
+            Statement::StructDef(_) => Some("struct_defn".to_string()),
+            Statement::EnumDef(_) => Some("enum_defn".to_string()),
+            Statement::TraitDef(_) => Some("trait_defn".to_string()),
+        };
 
-        match &self {
-            Statement::Expr(expr) => lines.extend(expr.node_display(next_indent)),
-            Statement::Return(expr) => {
-                lines.push("ret".to_string());
-
-                // if a return statement has a lin
-                if let Some(ret_expr) = expr {
-                    next_lines.push(format!(
-                        "{}{}",
-                        END_PIPE,
-                        ret_expr.node_display(next_indent).join("\n")
-                    ));
+        let child_lines: Vec<String> = match &self {
```

Comment:
Return tuple of `(node_name, child_lines)` from each branch, instead of having two separate `match`es